Parameterize FixedVector, EmbeddedFixedVector on Malloc type

Achierius opened the pull request:
#### b0c6df0dd951b0589cd74ac7baad087305fc8a1b
<pre>
Parameterize FixedVector, EmbeddedFixedVector on Malloc type
<a href="https://bugs.webkit.org/show_bug.cgi?id=286426">https://bugs.webkit.org/show_bug.cgi?id=286426</a>
<a href="https://rdar.apple.com/143497080">rdar://143497080</a>

Reviewed by Mark Lam.

This allows you to pass in a Malloc parameter to FixedVector,
which will be used to allocate the backing store.

* Source/WTF/wtf/EmbeddedFixedVector.h:
* Source/WTF/wtf/FixedVector.h:
(WTF::FixedVector::createWithSizeAndConstructorArguments):
(WTF::FixedVector::createWithSizeFromGenerator):
(WTF::FixedVector::operator== const):
(WTF::FixedVector::swap):
(WTF::Malloc&gt;::contains const):
(WTF::Malloc&gt;::findIf const):
(WTF::Malloc&gt;::find const):
(WTF::swap):
(WTF::map):
(WTF::FixedVector&lt;T&gt;::contains const): Deleted.
(WTF::FixedVector&lt;T&gt;::findIf const): Deleted.
(WTF::FixedVector&lt;T&gt;::find const): Deleted.
* Source/WTF/wtf/Forward.h:

Canonical link: <a href="https://commits.webkit.org/289364@main">https://commits.webkit.org/289364@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5a8e9a9879b1aa504ec9a78ced2462e976314eee

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/86686 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6193 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41018 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/91531 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/37418 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/6461 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/14251 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/67031 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/24817 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/89689 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/4917 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/78491 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47353 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/4712 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/32823 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/36536 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/79469 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/75159 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/33708 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/93426 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/85458 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/13839 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/10036 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/75830 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14040 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74314 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75017 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18454 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19328 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/17728 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/6624 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/13862 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/19122 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/107951 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/13600 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25980 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17044 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/15384 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/WTF/wtf/FixedVector.h:
```diff
@@ -29,10 +29,11 @@
 public:
-    using Storage = EmbeddedFixedVector<T>;
+    using Storage = EmbeddedFixedVector<T, Malloc>;
+    using Base = FixedVector<T, Malloc>;
```
Base means base class.  This is not the base class.  Calling it `Base` would be misleading.  Maybe call it `Self`?  There's no precedence for this.  In WebKit code, we typically spell it out, though that can get hairy when the template arguments are long.  Calling it `Self` or `This` might be ok.

Achierius replied:
I picked this up from VectorBuffer but I see that it's indeed using it for the proper base-class. Will change

MenloDorian commented on Source/WTF/wtf/FixedVector.h:
```diff
@@ -236,25 +237,25 @@ size_t FixedVector<T>::findIf(const MatchFunction& matches) const
 
-template<typename T, typename MapFunction, typename ReturnType = typename std::invoke_result<MapFunction, const T&>::type>
-FixedVector<ReturnType> map(const FixedVector<T>& source, MapFunction&& mapFunction)
+template<typename T, typename MapFunction, typename ReturnType = typename std::invoke_result<MapFunction, const T&>::type, typename Malloc = EmbeddedFixedVectorMalloc>
```
Is this default really necessary?  Would the compiler just infer it from the `const FixedVector<T, Malloc>& source` argument?

Achierius replied:
It's necessary since the prior template argument has a default, but it seems like nobody is using this right now so I can move the Malloc function to the third place so that it's before any defaulted template args.

MenloDorian commented on Source/WTF/wtf/Forward.h:
```diff
@@ -75,14 +75,15 @@ struct VectorBufferMalloc;
 #else
 using VectorBufferMalloc = FastMalloc;
 #endif
+using EmbeddedFixedVectorMalloc = FastMalloc;
```
This is not correct.  In EmbeddedFixedVector.h, `DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(EmbeddedFixedVector);` defines `EmbeddedFixedVectorMalloc`.  It is not always `FastMalloc`.

See how `VectorBufferMalloc` is handled above for the treatment you should apply.

Achierius replied:
Fixed

MenloDorian reviewed (approved):
r=me

Achierius committed:
Parameterize FixedVector, EmbeddedFixedVector on Malloc type
https://bugs.webkit.org/show_bug.cgi?id=286426
rdar://143497080

Reviewed by Mark Lam.

This allows you to pass in a Malloc parameter to FixedVector,
which will be used to allocate the backing store.

* Source/WTF/wtf/EmbeddedFixedVector.h:
* Source/WTF/wtf/FixedVector.h:
(WTF::FixedVector::createWithSizeAndConstructorArguments):
(WTF::FixedVector::createWithSizeFromGenerator):
(WTF::FixedVector::operator== const):
(WTF::FixedVector::swap):
(WTF::Malloc>::contains const):
(WTF::Malloc>::findIf const):
(WTF::Malloc>::find const):
(WTF::swap):
(WTF::map):
(WTF::FixedVector<T>::contains const): Deleted.
(WTF::FixedVector<T>::findIf const): Deleted.
(WTF::FixedVector<T>::find const): Deleted.
* Source/WTF/wtf/Forward.h:

Canonical link: https://commits.webkit.org/289364@main
